### PR TITLE
Feat/categories bulk actions

### PR DESCRIPTION
**Pull request by ArekChr**
### Details

This PR adds the ability to perform bulk actions on categories, such as disable, enable, or remove multiple categories at once.

image

### Fixed Issues

$ https://github.com/Expensify/App/issues/35709
PROPOSAL: https://docs.google.com/document/d/1gk3xqOs7epMbUrSSiX8K7YcqfPLVgqEos0sf-D-GMDA/edit#bookmark=id.onqkmvshcfhp

### Tests

- [ ] Verify that no errors appear in the JS console

1. Log in as an admin of a paid policy group and navigate to the workspace categories settings.
2. Choose some categories that are currently disabled.
3. Click the options menu in the top right corner and select "Enable Categories".
4. Confirm that the selected categories are now enabled.
5. Choose some categories that are currently enabled.
6. Click the options menu in the top right corner again and select "Disable Categories".
7. Check that these categories are now disabled.
8. Choose any categories you wish to delete.
9. In the top right corner, click the options menu and choose "Delete Categories".
10. Verify that the selected categories have been deleted.

### Network Failures Test

1. Log in as the admin of a paid policy group.
2. Go to profile settings, then navigate to "About" -> "Troubleshooting" and enable "Simulate Network Failures".
3. Go to the workspace categories settings.
4. Select some categories, whether they are currently disabled or enabled.
5. Use the options menu in the top right corner to either enable or disable the selected categories.
6. Wait for a moment, then look for an error message to appear under each category you tried to update, indicating a request failure.
7. Repeat steps 3 to 6 for the action of deleting categories.

### Offline tests

Here are the simplified and clarified test steps:

1. Log in as an admin of a paid policy group and open the workspace categories settings.
2. Disable your internet connection.
3. Choose some categories that are currently disabled.
4. Click the options menu in the top right corner and select "Enable Categories".
5. Check that the selected categories now appear as pending (shown in gray).
6. Choose some categories that are currently enabled.
7. Click the options menu in the top right corner again and select "Disable Categories".
8. Ensure that these categories now show as pending (in gray).
9. Select any categories you want to delete.
10. In the top right corner, click the options menu and choose "Delete Categories".
12. Confirm that the chosen categories are marked as pending (grayed out) and have a strikethrough.
13. Reconnect to the internet and verify that:
    - Categories that were pending disable are now enabled.
    - Categories that were pending enable are now disabled.
    - Categories marked for deletion are removed.

### QA Steps

The same steps as above.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1709744184746739
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/24796318/4c30060a-c81d-49f3-8db3-7137d5384d72

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/24796318/003b6364-6823-449a-b73e-a1fe00255c68

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/24796318/1e898511-c091-4186-bf49-1dfe12b7a68a

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/24796318/bdc978e4-a757-4668-bb50-7e68ccd792d8

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/24796318/67bd259d-faae-48f5-a405-2cf8fb9e647a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/24796318/ac32c9ae-3f38-476c-97ab-8cdffbafb865

</details>

### CODE REVIEW

File: assets/images/documentSlash.svg
Comment (rezkiy37) [inline assets/images/documentSlash.svg:1]: I see that we use a kind of dash naming of files for SVG icons like `document-slash.svg`.

File: src/libs/actions/Policy.ts [hunk 1]
```diff
@@ -2503,6 +2507,60 @@ function clearCategoryErrors(policyID: string, categoryName: string) {
...
+            },
+        ],
+    };
+
+    const parameters = {
+        policyID,
+        categories: JSON.stringify(categoryNamesToDelete),
+    };
+
+    API.write('DeleteWorkspaceCategories', parameters, onyxData);
```
Comment (rezkiy37) [inline src/libs/actions/Policy.ts:2561]: It should be a constant.
```suggestion
    API.write(WRITE_COMMANDS.DELETE_WORKSPACE_CATEGORIES, parameters, onyxData);
```

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 2]
```diff
@@ -50,6 +49,7 @@ function WorkspaceCategoriesPage({policyCategories, route}: WorkspaceCategoriesP
     const theme = useTheme();
     const {translate} = useLocalize();
     const [selectedCategories, setSelectedCategories] = useState<Record<string, boolean>>({});
+    const dropdownButtonRef = useRef(null);
```
Comment (rezkiy37) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:52]: The ref must be typed.
```suggestion
    const dropdownButtonRef = useRef<ButtonWithDropdownMenu | null>(null);
```

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 3]
```diff
@@ -62,38 +62,49 @@ function WorkspaceCategoriesPage({policyCategories, route}: WorkspaceCategoriesP
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
-    const categoryList = useMemo<PolicyForList[]>(
+    const categoryList = useMemo(
```
Comment (rezkiy37) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:65]:
```suggestion
    const categoryList = useMemo<PolicyOption>(
```

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 4]
```diff
@@ -103,27 +114,105 @@ function WorkspaceCategoriesPage({policyCategories, route}: WorkspaceCategoriesP
...
-                text={translate('common.settings')}
-                style={[isSmallScreenWidth && styles.w50]}
-            />
-        </View>
-    );
+    const getHeaderButtons = () => {
+        if (selectedCategoriesArray.length > 0) {
+            const isAllEnabled = selectedCategoriesArray.every((categoryName) => policyCategories?.[categoryName].enabled);
+
+            const options: Array<DropdownOption<DeepValueOf<typeof CONST.POLICY.CATEGORIES_BULK_ACTION_TYPES>>> = [
```
Comment (rezkiy37) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:135]: Looks like this type - `DeepValueOf<typeof CONST.POLICY.CATEGORIES_BULK_ACTION_TYPES>` - is reused a few times. Does it make sense to extract it to a separate type?

**Review by rezkiy37 — COMMENTED** [review]
Left comments.

**Comment by ArekChr** [comment]
@luacmartins There is an issue with the `openWorkspace` API call that seems to override pending actions for categories. Specifically, when a category is removed while offline, and then the page is reloaded or, on mobile, when navigating back and reopening the categories list, the action no longer appears as pending.

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 4 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:135]: Yes, but I think we can also remove `DeepValueOf<typeof CONST.POLICY.CATEGORIES_BULK_ACTION_TYPES>` from <ButtonWithDropdownMenu>

Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:135]: ButtonWithDropdownMenu accepts a generic type from options

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 3 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:65]: Is this necessary? We already have the PolicyOption type specified in the map function, which means the type of the category list is already defined.

Comment (rezkiy37) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:65]: Yeah, I saw it. I think that the typing of the `useMemo` is preferable. I would say it is better to type a kind of high-level function.

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 4 above)
Comment (rezkiy37) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:135]: Sorry, I am not sure I understand your idea.

Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:135]: check commit: 
[refactor(ts): remove generic from ButtonWithDropdownMenu](https://github.com/Expensify/App/pull/37734/commits/05ee7390fad6d65880f9947e01e89559d1d0853c)

File: src/languages/es.ts [hunk 5]
```diff
@@ -1785,6 +1786,10 @@ export default {
             collect: 'Recolectar',
         },
         categories: {
+            deleteCategories: 'Eliminar categorías',
+            disableCategories: 'Desactivar categorías',
+            enableCategories: 'Activar categorías',
+            deleteFailureMessage: 'Se ha producido un error al intentar eliminar la categoría. Por favor, inténtalo más tarde.',
```
Comment (akinwale) [inline src/languages/es.ts:1792]: Have the Spanish translations been verified?

**Review by akinwale — COMMENTED** [review]
Just an observation which is design-related, but it would be helpful to have some sort of visual differentiator between enabled / disabled categories so that I can tell at a glance what is enabled and what isn't. Right now, I have to look to the rhs of every category to be able to tell the difference.

File: src/languages/es.ts (hunk 5 above)
Comment (ArekChr) [reply src/languages/es.ts:1792]: not yet, sending translation to open source channel

Comment (luacmartins) [reply src/languages/es.ts:1792]: please tag me so I can assist

**Comment by luacmartins** [comment]
@ArekChr we have conflicts now

**Comment by ArekChr** [comment]
Conflicts resolved

File: src/languages/es.ts (hunk 5 above)
Comment (luacmartins) [reply src/languages/es.ts:1792]: Seems like these translations were aprpoved!

**Comment by luacmartins** [comment]
@ArekChr new conflicts

**Comment by luacmartins** [comment]
@akinwale let's continue the review, that way we can address all issues once @ArekChr starts his day tomorrow

File: src/libs/Navigation/types.ts [hunk 6]
```diff
@@ -520,7 +520,9 @@ type FullScreenNavigatorParamList = {
 type BottomTabNavigatorParamList = {
     [SCREENS.HOME]: undefined;
     [SCREENS.ALL_SETTINGS]: undefined;
-    [SCREENS.WORKSPACE.INITIAL]: undefined;
+    [SCREENS.WORKSPACE.INITIAL]: {
+        policyID: string;
+    };
```
Comment (luacmartins) [inline src/libs/Navigation/types.ts:525]: Why this change?

File: src/libs/actions/Policy.ts [hunk 7]
```diff
@@ -2378,6 +2379,7 @@ function setWorkspaceCategoryEnabled(policyID: string, categoriesToUpdate: Recor
                             pendingFields: {
                                 enabled: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
                             },
+                            pendingAction: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE,
                         };
```
Comment (luacmartins) [inline src/libs/actions/Policy.ts:2383]: I don't think we need both of these. We either define `pendingFields` or `pendingAction`. I think in this case we want `pendingField`

File: src/libs/actions/Policy.ts [hunk 8]
```diff
@@ -2398,6 +2400,7 @@ function setWorkspaceCategoryEnabled(policyID: string, categoriesToUpdate: Recor
                             pendingFields: {
                                 enabled: null,
                             },
+                            pendingAction: null,
```
Comment (luacmartins) [inline src/libs/actions/Policy.ts:2403]: same here and on other instances below

File: src/libs/actions/Policy.ts [hunk 9]
```diff
@@ -2503,6 +2507,60 @@ function clearCategoryErrors(policyID: string, categoryName: string) {
...
+            {
+                onyxMethod: Onyx.METHOD.SET,
+                key: `${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${policyID}`,
+                value: Object.keys(policyCategories).reduce<PolicyCategories>((acc, key) => {
+                    if (!categoryNamesToDelete.includes(key)) {
+                        acc[key] = policyCategories[key];
+                    }
+
+                    return acc;
+                }, {}),
```
Comment (luacmartins) [inline src/libs/actions/Policy.ts:2534]: I don't think we need to iterate over all categories, we could do something like this, no?

```suggestion
                value: categoryNamesToDelete.reduce<Record<string, Partial<PolicyCategory>>>((acc, categoryName) => {
                    acc[categoryName] = null;
                    return acc;
                }, {}),
```

File: src/libs/actions/Policy.ts [hunk 10]
```diff
@@ -2503,6 +2507,60 @@ function clearCategoryErrors(policyID: string, categoryName: string) {
...
+            },
+        ],
+        failureData: [
+            {
+                onyxMethod: Onyx.METHOD.MERGE,
+                key: `${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${policyID}`,
+                value: Object.keys(policyCategories).reduce<PolicyCategories>((acc, key) => {
```
Comment (luacmartins) [inline src/libs/actions/Policy.ts:2541]: Same here, we can just iterate over `categoryNamesToDelete` and remove pendingAction and add the error

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 11]
```diff
@@ -62,40 +62,61 @@ function WorkspaceCategoriesPage({policyCategories, route}: WorkspaceCategoriesP
...
             ...prev,
-            [category.value]: !prev[category.value],
+            [category.keyForList]: !prev[category.keyForList],
         }));
     };
 
     const toggleAllCategories = () => {
-        const isAllSelected = categoryList.every((category) => !!selectedCategories[category.value]);
-        setSelectedCategories(isAllSelected ? {} : Object.fromEntries(categoryList.map((item) => [item.value, true])));
+        const isAllSelected = categoryList.every((category) => !!selectedCategories[category.keyForList]);
```
Comment (luacmartins) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:114]: Maybe we could just compare the list size here, i.e. if `categoryList.length() === selectedCategories`. Not sure how much that would help with performance.

**Review by luacmartins — CHANGES_REQUESTED** [review]

File: assets/images/document-slash.svg
[the remark this answers is not in the document]
Comment (shawnborton) [reply assets/images/document-slash.svg:1]: Hmm I have no idea what is the correct asset. @trjExpensify you might have worked on this part of the design doc?

[the remark this answers is not in the document]
Comment (luacmartins) [reply assets/images/document-slash.svg:1]: Oh I meant that the icon doesn't seem to have the right size

[the remark this answers is not in the document]
Comment (dannymcclain) [reply assets/images/document-slash.svg:1]: It looks right to me? 🤷‍♂️

[the remark this answers is not in the document]
Comment (luacmartins) [reply assets/images/document-slash.svg:1]: Cool, just wanted to flag it in case it wasn't correct, but it seems like it is!

**Comment by akinwale** [comment]
@ArekChr I am not getting any errors when I enable the `Simulate Failing Network Requests` option and try to enable / disable categories.

https://github.com/Expensify/App/assets/4160319/34328291-94f1-4760-a1e5-21cdaa202a99

File: src/libs/actions/Policy.ts (hunk 9 above)
Comment (ArekChr) [reply src/libs/actions/Policy.ts:2534]: Sure! Fixing

**Comment by ArekChr** [comment]
> @ArekChr I am not getting any errors when I enable the Simulate Failing Network Requests option and try to enable / disable categories.

Could you please wait a bit longer? It attempts several times before indicating a failure.

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 11 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx]: Sure! refactored toggle category function to filter out categories with false values, to keep state more clean, please review

```
    const toggleCategory = (category: PolicyOption) => {
        setSelectedCategories((prev) => {
            if (prev[category.keyForList]) {
                const {[category.keyForList]: omittedCategory, ...newCategories} = prev;
                return newCategories;
            }
            return {...prev, [category.keyForList]: true};
        });
    };
    
   ```

**Comment by shawnborton** [comment]
I think in another PR somewhere we updated the enabled/disabled text to use a smaller font size and a slightly different color. I'm not sure if that PR was merged yet, but maybe a quick pull of main couldn't hurt?

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 12]
```diff
@@ -67,43 +67,63 @@ function WorkspaceCategoriesPage({policy, policyCategories, route}: WorkspaceCat
...
 
-    const toggleCategory = (category: PolicyForList) => {
-        setSelectedCategories((prev) => ({
-            ...prev,
-            [category.value]: !prev[category.value],
-        }));
+    const toggleCategory = (category: PolicyOption) => {
+        setSelectedCategories((prev) => {
+            if (prev[category.keyForList]) {
+                const {[category.keyForList]: omittedCategory, ...newCategories} = prev;
```
Comment (luacmartins) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:116]: What is `omittedCategory` and are we reusing it anywhere else?

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 13]
```diff
@@ -113,41 +133,119 @@ function WorkspaceCategoriesPage({policy, policyCategories, route}: WorkspaceCat
...
+                options.push({
+                    icon: Expensicons.DocumentSlash,
+                    text: translate('workspace.categories.disableCategories'),
+                    value: CONST.POLICY.CATEGORIES_BULK_ACTION_TYPES.DISABLE,
+                    onSelected: () => {
+                        setSelectedCategories({});
+                        setWorkspaceCategoryEnabled(route.params.policyID, categoriesToDisable);
+                    },
+                });
+            } else {
```
Comment (luacmartins) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:190]: I don't think this should be an if-else. If some categories are enabled and some are disabled, we want to show both options

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx [hunk 14]
```diff
@@ -113,41 +133,119 @@ function WorkspaceCategoriesPage({policy, policyCategories, route}: WorkspaceCat
...
+                    text: translate('workspace.categories.deleteCategories'),
+                    value: CONST.POLICY.CATEGORIES_BULK_ACTION_TYPES.DELETE,
+                    onSelected: () => {
+                        setSelectedCategories({});
+                        deleteWorkspaceCategories(route.params.policyID, selectedCategoriesArray);
+                    },
+                },
+            ];
+
+            if (isAllEnabled) {
```
Comment (luacmartins) [inline src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:170]: I don't think this is correct. We want to include this option is some categories are enabled, not all

**Review by luacmartins — CHANGES_REQUESTED** [review]

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 12 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:116]: I suggest using a destructuring pattern to exclude specific categories we don't want to include in the object. This approach allows us to keep only the selected categories in the list, rather than maintaining unselected categories with a value of false. WTYT?

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 13 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:190]: How should the functionality work if we select a mix of enabled and disabled categories? Should we have a feature that only enables the previously disabled categories, and another that only disables the previously enabled ones? What's your perspective on this?

Comment (luacmartins) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:190]: Yes, you can see how we're doing it for distance rates [here](https://github.com/Expensify/App/pull/37717/files#diff-bc07e9853f8d016b7882bd8bfc8c58153e0cee5b3b5f5f3ba0365ac20b59f8deR152-R183). If any selected category is enabled, we should add the `Disable` option, similar for disabled categories. We then filter the categories that don't need to change state before making the API call

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 12 above)
Comment (luacmartins) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:116]: What about something like this? https://github.com/Expensify/App/pull/37717/files#diff-bc07e9853f8d016b7882bd8bfc8c58153e0cee5b3b5f5f3ba0365ac20b59f8deR129-R135

Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:116]: @luacmartins I'm happy to bring back storing categories in the array. But you mentioned in first PR that we might have thousands of categories, iterating through objects probably might take in this case dozen milliseconds in low-end devices

**Comment by ArekChr** [comment]
@luacmartins It appears that the collection policy free trial period has expired. Looks like I found bug, should we fix this by hiding the category menu item? What is the expected behavior here?

image

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 13 above)
Comment (ArekChr) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:190]: When configuring options, if a category's status changes (e.g., from enabled to disabled), should we still include categories that remain unchanged in our selection? For instance, when mixing enabled and disabled categories, if we decide to disable the previously enabled ones, should we deselect all categories or only those whose statuses have been altered?

**Comment by ArekChr** [comment]
> I think in another PR somewhere we updated the enabled/disabled text to use a smaller font size and a slightly different color. I'm not sure if that PR was merged yet, but maybe a quick pull of main couldn't hurt?

Pulled main

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 12 above)
Comment (luacmartins) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:116]: Hmm that's a good point. I think we can leave the logic as is then.

File: src/pages/workspace/categories/WorkspaceCategoriesPage.tsx (hunk 13 above)
Comment (luacmartins) [reply src/pages/workspace/categories/WorkspaceCategoriesPage.tsx]: I think we should keep all of them selected, but the API call should filter out categories that are not changing state.

**Comment by luacmartins** [comment]
>@luacmartins It appears that the collect policy free trial period has expired. Looks like I found bug, should we fix this by hiding the category menu item? What is the expected behavior here?

I started seeing that in dev too, maybe there's a regression somewhere. Gonna investigate.

**Review by luacmartins — DISMISSED** [review]
LGTM and tests well!

**Comment by luacmartins** [comment]
@ArekChr we have conflicts again

**Comment by akinwale** [comment]
@ArekChr A couple things for dealing with network request failures.
1. The enabled / disabled state does not revert back to its previous value. Is this expected behaviour?
2. If a category or multiple categories are deleted, and the request fails, there is nothing to indicate that the operation failed. I assume this isn't as critical, but just want to confirm that this is also expected.

https://github.com/Expensify/App/assets/4160319/6189698a-f8de-4875-9ed7-0641593af217

**Comment by ArekChr** [comment]
> @ArekChr A couple things for dealing with network request failures.
> 
> 1. The enabled / disabled state does not revert back to its previous value. Is this expected behaviour?
> 2. If a category or multiple categories are deleted, and the request fails, there is nothing to indicate that the operation failed. I assume this isn't as critical, but just want to confirm that this is also expected.
> 
>  37734-web-network-failure.mp4

@akinwale I encountered the same issue again and after re-logging in, the updated state is still visible despite receiving an error message. It seems the state updates are going through even when we encounter simulation failure requests. Should we consider filing a bug report on this matter? @lucamartins, have you dealt with something similar in your experience?

**Comment by luacmartins** [comment]
>@lucamartins, have you dealt with something similar in your experience?

TBH I've never gotten the `Simulate failing network requests` feature to work reliably. I always throw an error on our dev API to test these. I think we can move forward without blocking on that.

**Review by luacmartins — APPROVED** [review]

**Review by akinwale — APPROVED** [review]
LGTM.

**Review by luacmartins — DISMISSED** [review]

**Comment by luacmartins** [comment]
@ArekChr more conflicts and TS is failing again

**Review by luacmartins — APPROVED** [review]

**Comment by luacmartins** [comment]
Tests passed when I merged it

**Comment by cubuspl42** [comment]
Coming from https://github.com/Expensify/App/issues/38262

When a category is disabled while offline, the expected behavior was to make the entire category row grayed out. Only the status column was grayed out.
